crates/binjs_io/src/entropy/predict.rs

```diff
@@ -336,3 +346,172 @@ impl<NodeValue> PathPredict<NodeValue, SymbolInfo> where NodeValue: Eq + Hash +
+    value_by_dictionary_index: Vec<NodeValue>,
+    dictionary_index_by_value: HashMap<NodeValue, DictionaryIndex>,
+
+    latest_values: Vec<NodeValue>,
+
+    info: ContextInformation<WindowPrediction, Statistics>,
+}
+impl<NodeValue, Statistics> WindowPredict<NodeValue, Statistics> where NodeValue: Clone + Eq + Hash {
+    pub fn new(width: usize) -> Self {
+        WindowPredict {
+            width,
+            value_by_dictionary_index: Vec::with_capacity(1024), // FIXME: Magic number
+            dictionary_index_by_value: HashMap::with_capacity(1024),
+            latest_values: Vec::with_capacity(width),
+            info: ContextInformation::new(),
+        }
+    }
+
+    /// Update current window by moving an index to the first position.
+    fn update_window_by_index(&mut self, index: BackReference) -> Result<BackReference, ()> {
+        let as_usize: usize = index.into();
+        if as_usize >= self.latest_values.len() {
+            return Err(());
+        }
+        let ref mut slice = self.latest_values
+            .as_mut_slice()
+            [0..index.into()];
+        slice.rotate_right(1);
+        Ok(index)
+    }
+
+    /// Update current window by putting a value to the first position.
+    ///
+    /// If the value is already in the window, it is moved to the first position,
+    /// otherwise, it is added. If the resulting window is too large, it is truncated.
+    fn update_window_by_value(&mut self, value: &NodeValue) -> Option<BackReference> {
+        // It's possible that the value was present in the window.
+        if let Some(index) = self.latest_values
+            .iter()
+            .position(|v| v == value)
+        {
+            return Some(self.update_window_by_index(BackReference(index))
+                .unwrap())
+        }
+        if self.latest_values.len() < self.width {
+            // If the window isn't full yet, simply add the value at start.
+            self.latest_values.insert(0, value.clone());
+        } else {
+            // Otherwise, rotate.
+            let slice = self.latest_values
+                .as_mut_slice();
+            slice[slice.len() - 1] = value.clone();
+            slice.rotate_right(1);
+        }
+        None
+    }
+}
+impl<NodeValue> WindowPredict<NodeValue, Instances> where NodeValue: Clone + Eq + std::hash::Hash + std::fmt::Debug {
+    pub fn add(&mut self, value: NodeValue) {
+        // --- At this stage, we don't know whether the value is known.
+
+        let number_of_values = self.value_by_dictionary_index.len();
+        let dictionary_index = *self.dictionary_index_by_value.entry(value.clone())
+            .or_insert(DictionaryIndex(number_of_values));
+
+        if dictionary_index == DictionaryIndex(number_of_values) {
+            // We've just inserted `value`.
+            self.value_by_dictionary_index.push(value.clone());
+        } else {
+            // Value was already known.
+            let index : usize = dictionary_index.into();
```

Comment:
won't this hit unused variable on non-debug build?
(I don't know much about debug/non-debug build difference and warning system tho)

Reply:
@arai-a I'm almost sure that it won't, but I'll admit that I haven't checked.